### PR TITLE
Figure out some edge cases

### PR DESCRIPTION
**Pull request by ibizaman**
Tries some edge cases and see if we need to tweak the interface definition to accommodate them:

- Use directly `input`, reducing boilerplate.
- Add provider specific option.
- Uses provider option type inside an `attrsOf`.

---

Previous description:

> Compared to the string example, it:
> 
> - Uses attrsOf to create multiple provider instances.
> - Adds provider specific options with values different for each instance.
> 
> This is WIP because I'd like to create a NixOS test for this so I can actually use `pkgs.writeText`. Actually, it could work without using NixOS tests as long as we import the correct modules. Also, I'd like to add a provider for sops, not sure yet how to do that.
> 
>

### CODE REVIEW

File: examples/secrets/secrets-teststub-provider.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+let
+  inherit (lib) nameValuePair mapAttrs' mkOption types;
+  # inherit (pkgs) writeText;
+in
+{
+  options.secrets-teststub-provider = mkOption {
+    type = with types; submodule ({
+      options = {
+        secrets = mkOption {
+          type = types.attrsOf (submodule (providerSpecific@{ name, ... }: {
```
Comment (fricklerhandwerk) [inline examples/secrets/secrets-teststub-provider.nix]: Ah, this is clever, passing provider-specific options per provider instance. I like it!

File: default.nix [hunk 2]
```diff
@@ -23,28 +23,35 @@ in
...
-            ${getExe watchexec} -w ${toString ./.} -- nix-instantiate --eval --strict ./. -A example.config.string-consumer.output --json "$@"
+            ${getExe watchexec} -w ${toString ./.} -- nix-instantiate --eval --strict ./. -A ${what} --json "$@"
           '';
       };
     in
     {
-      shell = pkgs.mkShellNoCC {
+      shell = pkgs.mkShellNoCC (let
+          test-loop-string = test-loop "test-loop-string" "example.config.string-consumer.output";
+          test-loop-secrets = test-loop "test-loop-secrets" "example.config.secrets-consumer.password.output";
```
Comment (fricklerhandwerk) [inline default.nix]: Instead, as a preliminary we should switch to nix-unit (or the Nixpkgs based unit testing if you want it to be easier to upstream) and simply test all samples in a loop.

File: examples/secrets/secrets-teststub-provider.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+              };
+            };
+          }));
+        };
+      };
+    });
+  };
+
+  config = {
+    # system.activationScripts = mapAttrs' (n: cfg':
```
Comment (fricklerhandwerk) [inline examples/secrets/secrets-teststub-provider.nix]: I think it's not necessary to do this in order to show the principle. If anything, to demo generates secrets at activation time, we can do a separate PR.

Comment (fricklerhandwerk) [reply examples/secrets/secrets-teststub-provider.nix]: Arguably even, to emphasize the principle of parameterised providers and avoid distractions from pre-existing knowledge, we may want to remove any semblance of practical examples. Essentially the string-provider-but-with-letter-a we had sketched.

**Review by fricklerhandwerk — COMMENTED** [review]
It struck me that we actually re-invented function types (i.e. signatures), not interfaces. Which the module system notably doesn't have! We only have half of them with `functionTo`. Interfaces (or type classes) would actually be sets of (parameterised) function types.

Interestingly

```
something.input = "bar";
something.provider = config.providers.foo;
result = config.something.output;
```

is a very cumbersome, modularly, but type safe way of saying `result = foo "bar"`.

So we may as well use names that don't obscure that:

```
apply.arg = "bar";
apply.function = config.functions.foo;
result = config.apply.value;
```

Observe how `config.functions` is now a namespace for functions (they can't be anonymous!), and even each function application needs to be manually named. Making the functions namespace an attrset is something akin to currying because it allows passing additional parameters.

This sure is geeky to an entertaining degree but I wonder if there's a simpler way to achieve type safe function application in the module system. 

(Also note how we can't garbage collect anything because the option tree stays around until everything has evaluated.)

**Comment by fricklerhandwerk** [comment]
```
{ config, ...}:
{
  baz = {
    type = config.qux;
    arg = config.bar;
    func = config.foo;
  };
}
```

I added a functional programming language to your functional programming language so you can have some functional programming in your functional programming.

File: examples/secrets/secrets-teststub-provider.nix (hunk 3 above)
Comment (ibizaman) [reply examples/secrets/secrets-teststub-provider.nix]: Ah yes I agree. If you don’t mind, I’d like to add an example for all the middle interfaces I already had in my project as a way to stress test the current implementation. Then when we are sure it is sound, we can remove them but at least we’ll know we thought about most edge cases.

File: examples/secrets/secrets-teststub-provider.nix (hunk 1 above)
Comment (ibizaman) [reply examples/secrets/secrets-teststub-provider.nix]: That’s even more necessary for backups to be able to backup a same service to multiple repos.

**Comment by ibizaman** [comment]
I fully agree we're doing functions expressed as modules. I still think it's worth it, written this way, because we're not introducing this just for the sake of type funkiness. We come from a concrete background with a handful of diverse examples for which we wrote what make sense to us without knowing necessarily where we were getting at. And everything leads to needing this.

I also think the final syntax the end user needs to deal with is quite slick.

About the garbage collection, thats really an issue of the module system itself. Wouldn’t it be bettered solved, or at least tackled, through the work to split the monolithic all-modules?
